File: billing-v3/src/test/java/io/appmetrica/analytics/billingv3/impl/library/PurchaseHistoryResponseListenerImplTest.java
package io.appmetrica.analytics.billingv3.impl.library;

import androidx.annotation.NonNull;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.PurchaseHistoryRecord;
import com.android.billingclient.api.SkuDetailsParams;
import com.android.billingclient.api.SkuDetailsResponseListener;
import io.appmetrica.analytics.billinginterface.internal.BillingInfo;
import io.appmetrica.analytics.billinginterface.internal.ProductType;
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig;
import io.appmetrica.analytics.billinginterface.internal.library.UtilsProvider;
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoManager;
import io.appmetrica.analytics.billinginterface.internal.update.UpdatePolicy;
import io.appmetrica.analytics.coreutils.internal.executors.SafeRunnable;
import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class PurchaseHistoryResponseListenerImplTest {

    @Mock
    private Executor workerExecutor;
    @Mock
    private Executor uiExecutor;
    @Mock
    private BillingClient billingClient;
    @Mock
    private UtilsProvider utilsProvider;
    @Mock
    private UpdatePolicy updatePolicy;
    @Mock
    private BillingInfoManager billingInfoManager;
    @Mock
    private BillingLibraryConnectionHolder billingLibraryConnectionHolder;
    @Mock
    private SystemTimeProvider systemTimeProvider;

    private PurchaseHistoryResponseListenerImpl purchaseHistoryResponseListener;
    private final long now = 4142;
    private final BillingConfig billingConfig = new BillingConfig(41, 42);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                ((Runnable) invocation.getArgument(0)).run();
                return null;
            }
        }).when(workerExecutor).execute(any(SafeRunnable.class));
        doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                ((Runnable) invocation.getArgument(0)).run();
                return null;
            }
        }).when(uiExecutor).execute(any(SafeRunnable.class));
        when(utilsProvider.getUpdatePolicy()).thenReturn(updatePolicy);
        when(utilsProvider.getBillingInfoManager()).thenReturn(billingInfoManager);
        when(systemTimeProvider.currentTimeMillis()).thenReturn(now);

        purchaseHistoryResponseListener = new PurchaseHistoryResponseListenerImpl(billingConfig, workerExecutor, uiExecutor, billingClient, utilsProvider, BillingClient.SkuType.SUBS, billingLibraryConnectionHolder, systemTimeProvider);
    }

    @Test
    public void testOnPurchaseHistoryResponseIfError() {
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.ERROR)
                        .build(),
                null
        );
        verifyNoMoreInteractions(billingClient);
        verifyNoMoreInteractions(billingInfoManager);
        verify(billingLibraryConnectionHolder).removeListener(purchaseHistoryResponseListener);
    }

    @Test
    public void testOnPurchaseHistoryResponseIfNullList() {
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.OK)
                        .build(),
                null
        );
        verifyNoMoreInteractions(billingClient);
        verifyNoMoreInteractions(billingInfoManager);
        verify(billingLibraryConnectionHolder).removeListener(purchaseHistoryResponseListener);
    }

    @Test
    public void testOnPurchaseHistoryResponseIfOk() throws JSONException {
        when(billingClient.isReady()).thenReturn(true);
        when(updatePolicy.getBillingInfoToUpdate(
                any(BillingConfig.class),
                ArgumentMatchers.<String, BillingInfo>anyMap(),
                any(BillingInfoManager.class)
        )).thenReturn(getBillingInfoToUpdate(ProductType.SUBS));

        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.OK)
                        .build(),
                getPurchaseHistory()
        );
        ArgumentCaptor<Map<String, BillingInfo>> argument = ArgumentCaptor.forClass(Map.class);
        verify(updatePolicy).getBillingInfoToUpdate(eq(billingConfig), argument.capture(), eq(billingInfoManager));
        assertThat(argument.getValue()).usingRecursiveComparison().isEqualTo(getBillingInfoToUpdate(ProductType.SUBS));

        verifyNoMoreInteractions(billingInfoManager);
        verify(billingLibraryConnectionHolder).addListener(any(SkuDetailsResponseListenerImpl.class));
        verify(billingLibraryConnectionHolder, never()).removeListener(any(SkuDetailsResponseListenerImpl.class));
        verify(billingLibraryConnectionHolder).removeListener(purchaseHistoryResponseListener);
        verify(billingClient).querySkuDetailsAsync(any(SkuDetailsParams.class), any(SkuDetailsResponseListener.class));
    }

    @Test
    public void testUpdateStorageForNonEmptyNewBillingInfo() {
        when(billingInfoManager.isFirstInappCheckOccurred()).thenReturn(true);
        final Map<String, BillingInfo> history = new HashMap<>();
        history.put("sku1", new BillingInfo(ProductType.INAPP, "sku1", "token1", 10, 43));
        history.put("sku2", new BillingInfo(ProductType.INAPP, "sku2", "token2", 10, 42));

        final Map<String, BillingInfo> newBillingInfo = new HashMap<>();
        newBillingInfo.put("sku1", new BillingInfo(ProductType.INAPP, "sku1", "token1", 10, 0));

        purchaseHistoryResponseListener.updateStorage(history, newBillingInfo);

        final Map<String, BillingInfo> changedHistory = new HashMap<>();
        changedHistory.put("sku1", new BillingInfo(ProductType.INAPP, "sku1", "token1", 10, now));
        changedHistory.put("sku2", new BillingInfo(ProductType.INAPP, "sku2", "token2", 10, 42));

        ArgumentCaptor<Map<String, BillingInfo>> captor = ArgumentCaptor.forClass(Map.class);
        verify(billingInfoManager).update(captor.capture());
        assertThat(captor.getValue()).usingRecursiveComparison().isEqualTo(changedHistory);
    }

    @Test
    public void testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckOccurred() throws JSONException {
        when(billingClient.isReady()).thenReturn(true);
        when(updatePolicy.getBillingInfoToUpdate(
                any(BillingConfig.class),
                ArgumentMatchers.<String, BillingInfo>anyMap(),
                any(BillingInfoManager.class)
        )).thenReturn(new HashMap<String, BillingInfo>());
        when(billingInfoManager.isFirstInappCheckOccurred()).thenReturn(true);
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.OK)
                        .build(),
                getPurchaseHistory()
        );
        ArgumentCaptor<Map<String, BillingInfo>> captor = ArgumentCaptor.forClass(Map.class);
        verify(billingInfoManager).update(captor.capture());
        assertThat(captor.getValue()).usingRecursiveComparison().isEqualTo(getBillingInfoToUpdate(ProductType.SUBS));
        verify(billingInfoManager, never()).markFirstInappCheckOccurred();
    }

    @Test
    public void testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckNotOccurredAndTypeInapp() throws JSONException {
        purchaseHistoryResponseListener = new PurchaseHistoryResponseListenerImpl(billingConfig, workerExecutor, uiExecutor, billingClient, utilsProvider, BillingClient.SkuType.INAPP, billingLibraryConnectionHolder, systemTimeProvider);
        when(billingClient.isReady()).thenReturn(true);
        when(updatePolicy.getBillingInfoToUpdate(
                any(BillingConfig.class),
                ArgumentMatchers.<String, BillingInfo>anyMap(),
                any(BillingInfoManager.class)
        )).thenReturn(new HashMap<String, BillingInfo>());
        when(billingInfoManager.isFirstInappCheckOccurred()).thenReturn(false);
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.OK)
                        .build(),
                getPurchaseHistory()
        );
        ArgumentCaptor<Map<String, BillingInfo>> captor = ArgumentCaptor.forClass(Map.class);
        verify(billingInfoManager).update(captor.capture());
        assertThat(captor.getValue()).usingRecursiveComparison().isEqualTo(getBillingInfoToUpdate(ProductType.INAPP));
        verify(billingInfoManager).markFirstInappCheckOccurred();
    }

    @Test
    public void testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckNotOccurredAndTypeNotInapp() throws JSONException {
        purchaseHistoryResponseListener = new PurchaseHistoryResponseListenerImpl(billingConfig, workerExecutor, uiExecutor, billingClient, utilsProvider, BillingClient.SkuType.SUBS, billingLibraryConnectionHolder, systemTimeProvider);
        when(billingClient.isReady()).thenReturn(true);
        when(updatePolicy.getBillingInfoToUpdate(
                any(BillingConfig.class),
                ArgumentMatchers.<String, BillingInfo>anyMap(),
                any(BillingInfoManager.class)
        )).thenReturn(new HashMap<String, BillingInfo>());
        when(billingInfoManager.isFirstInappCheckOccurred()).thenReturn(false);
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
                BillingResult.newBuilder()
                        .setResponseCode(BillingClient.BillingResponseCode.OK)
                        .build(),
                getPurchaseHistory()
        );
        ArgumentCaptor<Map<String, BillingInfo>> captor = ArgumentCaptor.forClass(Map.class);
        verify(billingInfoManager).update(captor.capture());
        assertThat(captor.getValue()).usingRecursiveComparison().isEqualTo(getBillingInfoToUpdate(ProductType.SUBS));
        verify(billingInfoManager, never()).markFirstInappCheckOccurred();
    }

    private List<PurchaseHistoryRecord> getPurchaseHistory() throws JSONException {
        final List<PurchaseHistoryRecord> purchaseHistoryRecords = new ArrayList<>();
        purchaseHistoryRecords.add(new PurchaseHistoryRecord(
                new JSONObject()
                        .put("productId", "sku1")
                        .put("purchaseToken", "token1")
                        .put("purchaseTime", 10)
                        .toString(),
                "signature"
        ));
        return purchaseHistoryRecords;
    }

    private Map<String, BillingInfo> getBillingInfoToUpdate(@NonNull final ProductType type) {
        final Map<String, BillingInfo> result = new HashMap<>();
        result.put("sku1", new BillingInfo(type, "sku1", "token1", 10, 0));
        return result;
    }
}

File: billing-v4/src/test/java/io/appmetrica/analytics/billingv4/impl/library/PurchaseHistoryResponseListenerImplTest.kt
package io.appmetrica.analytics.billingv4.impl.library

import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.PurchaseHistoryRecord
import io.appmetrica.analytics.billinginterface.internal.BillingInfo
import io.appmetrica.analytics.billinginterface.internal.ProductType
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig
import io.appmetrica.analytics.billinginterface.internal.library.UtilsProvider
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoManager
import io.appmetrica.analytics.billinginterface.internal.update.UpdatePolicy
import io.appmetrica.analytics.coreutils.internal.executors.SafeRunnable
import org.assertj.core.api.Assertions.assertThat
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyMap
import org.mockito.Mock
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoInteractions
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.eq
import org.robolectric.RobolectricTestRunner
import java.util.concurrent.Executor

@RunWith(RobolectricTestRunner::class)
class PurchaseHistoryResponseListenerImplTest {

    @Mock
    private lateinit var workerExecutor: Executor
    @Mock
    private lateinit var uiExecutor: Executor
    @Mock
    private lateinit var billingClient: BillingClient
    @Mock
    private lateinit var utilsProvider: UtilsProvider
    @Mock
    private lateinit var updatePolicy: UpdatePolicy
    @Mock
    private lateinit var billingInfoManager: BillingInfoManager
    @Mock
    private lateinit var billingLibraryConnectionHolder: BillingLibraryConnectionHolder

    private lateinit var purchaseHistoryResponseListener: PurchaseHistoryResponseListenerImpl
    private val billingConfig = BillingConfig(41, 42)

    @Before
    fun setUp() {
        MockitoAnnotations.openMocks(this)
        doAnswer { invocation ->
            (invocation.getArgument<Any>(0) as Runnable).run()
            null
        }.`when`(workerExecutor).execute(any<SafeRunnable>())
        doAnswer { invocation ->
            (invocation.getArgument<Any>(0) as Runnable).run()
            null
        }.`when`(uiExecutor).execute(any<SafeRunnable>())
        `when`(utilsProvider.updatePolicy).thenReturn(updatePolicy)
        `when`(utilsProvider.billingInfoManager).thenReturn(billingInfoManager)
        `when`(utilsProvider.workerExecutor).thenReturn(workerExecutor)
        `when`(utilsProvider.uiExecutor).thenReturn(uiExecutor)

        purchaseHistoryResponseListener = PurchaseHistoryResponseListenerImpl(
            billingConfig,
            billingClient,
            utilsProvider,
            BillingClient.SkuType.SUBS,
            billingLibraryConnectionHolder
        )
    }

    @Test
    fun testOnPurchaseHistoryResponseIfError() {
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.ERROR)
                .build(),
            null
        )
        verifyNoInteractions(billingClient)
        verifyNoInteractions(billingInfoManager)
        verify(billingLibraryConnectionHolder).removeListener(purchaseHistoryResponseListener)
    }

    @Test
    fun testOnPurchaseHistoryResponseIfNullList() {
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.OK)
                .build(),
            null
        )
        verifyNoInteractions(billingClient)
        verifyNoInteractions(billingInfoManager)
        verify(billingLibraryConnectionHolder).removeListener(purchaseHistoryResponseListener)
    }

    @Test
    fun testOnPurchaseHistoryResponseIfOk() {
        `when`(billingClient.isReady).thenReturn(true)
        `when`(updatePolicy.getBillingInfoToUpdate(any(), anyMap(), any()))
            .thenReturn(getBillingInfoToUpdate(ProductType.SUBS))
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.OK)
                .build(),
            getPurchaseHistory()
        )
        val captor = argumentCaptor<Map<String, BillingInfo>>()

        verify(updatePolicy).getBillingInfoToUpdate(
            eq(billingConfig),
            captor.capture(),
            eq(billingInfoManager)
        )
        assertThat(captor.firstValue).usingRecursiveComparison().isEqualTo(
            getBillingInfoToUpdate(ProductType.SUBS)
        )
        verifyNoInteractions(billingInfoManager)
        verify(billingLibraryConnectionHolder).addListener(
            any<SkuDetailsResponseListenerImpl>()
        )
        verify(billingLibraryConnectionHolder, never()).removeListener(
            any<SkuDetailsResponseListenerImpl>()
        )
        verify(billingLibraryConnectionHolder).removeListener(
            purchaseHistoryResponseListener
        )
        verify(billingClient).querySkuDetailsAsync(
            any(),
            any()
        )
    }

    @Test
    fun testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckOccurred() {
        `when`(billingClient.isReady).thenReturn(true)
        `when`(updatePolicy.getBillingInfoToUpdate(any(), anyMap(), any())).thenReturn(mapOf())
        `when`(billingInfoManager.isFirstInappCheckOccurred).thenReturn(true)
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.OK)
                .build(),
            getPurchaseHistory()
        )
        val captor = argumentCaptor<Map<String, BillingInfo>>()
        verify(billingInfoManager).update(captor.capture())
        assertThat(captor.firstValue).usingRecursiveComparison().isEqualTo(
            getBillingInfoToUpdate(ProductType.SUBS)
        )
        verify(billingInfoManager, never()).markFirstInappCheckOccurred()
    }

    @Test
    fun testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckNotOccurredAndTypeInapp() {
        purchaseHistoryResponseListener = PurchaseHistoryResponseListenerImpl(
            billingConfig,
            billingClient,
            utilsProvider,
            BillingClient.SkuType.INAPP,
            billingLibraryConnectionHolder
        )
        `when`(billingClient.isReady).thenReturn(true)
        `when`(updatePolicy.getBillingInfoToUpdate(any(), anyMap(), any())).thenReturn(mapOf())
        `when`(billingInfoManager.isFirstInappCheckOccurred).thenReturn(false)
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.OK)
                .build(),
            getPurchaseHistory()
        )
        val captor = argumentCaptor<Map<String, BillingInfo>>()
        verify(billingInfoManager).update(captor.capture())
        assertThat(captor.firstValue).usingRecursiveComparison().isEqualTo(
            getBillingInfoToUpdate(ProductType.INAPP)
        )
        verify(billingInfoManager).markFirstInappCheckOccurred()
    }

    @Test
    fun testOnPurchaseHistoryResponseIfOkAndUpdatePolicyReturnsEmptyMapAndFirstInappCheckNotOccurredAndTypeNotInapp() {
        purchaseHistoryResponseListener = PurchaseHistoryResponseListenerImpl(
            billingConfig,
            billingClient,
            utilsProvider,
            BillingClient.SkuType.SUBS,
            billingLibraryConnectionHolder
        )
        `when`(billingClient.isReady).thenReturn(true)
        `when`(updatePolicy.getBillingInfoToUpdate(any(), anyMap(), any())).thenReturn(mapOf())
        `when`(billingInfoManager.isFirstInappCheckOccurred).thenReturn(false)
        purchaseHistoryResponseListener.onPurchaseHistoryResponse(
            BillingResult.newBuilder()
                .setResponseCode(BillingClient.BillingResponseCode.OK)
                .build(),
            getPurchaseHistory()
        )
        val captor = argumentCaptor<Map<String, BillingInfo>>()
        verify(billingInfoManager).update(captor.capture())
        assertThat(captor.firstValue).usingRecursiveComparison().isEqualTo(
            getBillingInfoToUpdate(ProductType.SUBS)
        )
        verify(billingInfoManager, never()).markFirstInappCheckOccurred()
    }

    private fun getPurchaseHistory(): List<PurchaseHistoryRecord> {
        val purchaseHistoryRecords: MutableList<PurchaseHistoryRecord> = ArrayList()
        purchaseHistoryRecords.add(
            PurchaseHistoryRecord(
                JSONObject()
                    .put("productId", "sku1")
                    .put("purchaseToken", "token1")
                    .put("purchaseTime", 10)
                    .toString(),
                "signature"
            )
        )
        return purchaseHistoryRecords
    }

    private fun getBillingInfoToUpdate(type: ProductType): Map<String, BillingInfo> {
        return mapOf(
            "sku1" to BillingInfo(type, "sku1", "token1", 10, 0)
        )
    }
}
